### PR TITLE
Implement Kakoune's A-S: Select first and last chars

### LINKED ISSUE #9459: Implement Kakoune's `A-S`
In Kakoune, selects the first and last characters of each selection.

`[helix]` becomes `[h]eli[x]`

### PR DESCRIPTION
**Pull request by scdailey**
Fixes #9459

### CODE REVIEW

**Comment by kirawi** [comment]
I don't think this handles zero-size selections (only when doc is completely empty) or 1-wide selections. Might want to check the size first.

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -1811,6 +1812,20 @@ fn select_all(cx: &mut Context) {
     doc.set_selection(view.id, Selection::single(0, end))
 }
 
+fn select_first_and_last_chars(cx: &mut Context) {
+    let (view, doc) = current!(cx.editor);
+
+    let selection = doc.selection(view.id).clone().transform_iter(|range| {
+        vec![
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:1819]:
```suggestion
        [
```

no need to create a Vec, we can iterate on a slice instead

**Comment by scdailey** [comment]
> I don't think this handles zero-size selections (only when doc is completely empty) or 1-wide selections. Might want to check the size first.

Good catch - I'll do that now

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -1811,6 +1812,20 @@ fn select_all(cx: &mut Context) {
...
+            Range::new(range.anchor, range.anchor + 1),
+            Range::new(range.head - 1, range.head),
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:1821]: The subtraction here can cause a crash: open a scratch buffer and delete the trailing newline so the buffer is empty, then `A-S` subtracts with overflow.

Instead let's use points and the prev_grapheme_boundary helper:

```suggestion
            Range::new(range.from()),
            Range::new(graphemes::prev_grapheme_boundary(text, range.to())),
```

**Review by the-mikedavis — CHANGES_REQUESTED** [review]
This should also be added to the book under selection manipulation: https://github.com/helix-editor/helix/blob/master/book/src/keymap.md#selection-manipulation

**Comment by scdailey** [comment]
Okay just made the fixes. I believe that using the helpers also addressed the overflow issues that @kirawi mentioned as well.

File: book/src/keymap.md
[the remark this answers is not in the document]
Comment (scdailey) [reply book/src/keymap.md:1]: Yeah sorry about that, I opened it in a markdown editor and I think it auto formatted it. Ill fix that along with the Range::point change

[the remark this answers is not in the document]
Comment (scdailey) [reply book/src/keymap.md:1]: Okay should be good now

**Pull request closed without merging**